### ISSUE TITLE
[Feature Request] Help certificate rotation of client certificates

### ISSUE THREAD

**Issue by jmprieur** [AzureAD/microsoft-identity-web#956]
**Is your feature request related to a problem? Please describe.**
Today, Microsoft.Identity.Web:
- gets certificates from KeyVault
- helps rotating decrypt certificates (by enabling passing several certificates)

For the client certificates, it also supports send5xc, however, it could help a bit more by fetching from keyvault a newer version of a certificate when the confidential client flow fail because of certificate expiry or invalidation.

**Describe the solution you'd like**
When:
- the certificate proposed by Microsoft.Identity.Web to prove the app identity is rejected by Azure AD (for instance because it was revoked in the app registration, or a newer version is available in KeyVault), Microsoft.Identity.Web should try to refetch the certificate from KeyVault.

**Describe alternatives you've considered**
Subscribe to KeyVault events, but this would require too much configuration from customers

**Comment by jmprieur** [comment]
### What happens: 
1. Create a self signed certificate which will expire soon. For instance if I want the cert to expire today at 4:05 pm. The date/time needs to be adapted to your locale (you might need to wrote "03/15/2021 4:05 pm" in the US

  ```PowerShell
   $cert=New-SelfSignedCertificate -Subject "Expired" -CertStoreLocation "Cert:\CurrentUser\My"  -KeyExportPolicy Exportable -KeySpec Signature -NotAfter "15/03/2021 16:05"
  ```

2. Using the certificate manager, export with private key

3. Add the certificate to the client certificate of the app to test before the certificate expires.

4. After the certificate expires, run the app, and observe that there is an MsalUiRequiredException:
   ` StatusCode: 401
        ResponseBody: {"error":"invalid_client","error_description":"AADSTS700027: Client assertion contains an invalid signature. [Reason - The key was not found., Thumbprint of key used by client: '24C1D15E417EDB6AA9DB1BCF05B3B00479809726'`

**Comment by wuwei8372** [comment]
A few questions:
1) Looks like proposed method will only re-load the certificate when the current certificate is expired or invalid. If a new version of certificate is available and the old version is still valid, could we get the new certificate when calling LoadIfNeeded? (Probably not, right?)

2) Will this proposed method also work for non-aad certificate? For example a client certificate used for service to service call?

3) Could we possibly still check the certificate.Properties.NotBefore after line 117 in  DefaultCertificateLoader.cs even CertificateDescription.Certificate is not null when calling LoadIfNeeded? Then if MIW finds the newest certificate from AKV has a newer expire date than the current CertificateDescription.Certificate, it can replace CertificateDescription.Certificate with the newer version from AKV. Could it be way to roll forward to the new cert even the current CertificateDescription.Certificate is still valid?

Thanks!

**Comment by jmprieur** [comment]
Thanks for these questions, @wuwei8372 

1. Would you think that we'd want to re-check the certificate periodically?
2. You can use the DefaultCertificateLoader to load certificates.
3. I think that this is related to question 1). I'd think that, in that case, we'd want to also keep a LKG (Last known good) certificate in case Azure AD's application was not yet updated ?

**Comment by wuwei8372** [comment]
Thanks for replay @jmprieur!

1. It would be great if we could re-check the and update certificateDescription.Certificate periodically to make it up-to-date.
2. Good to know we could use DefaultCertificateLoader load client certificates for S2S call as well.
3. Yes it is related to question 1. I assume LKG means the latest valid certificate in AKV, correct? If we can periodically check and put the latest certificate from AKV in certificateDescription.Certificate, then the App referring to certificateDescription.Certificate would always use the lasted cert from AKV, which will be great.

**Comment by jmprieur** [comment]
Yes @wuwei8372, LKG = Last Known Good (the latest valid certificate in AKV). The issue is, the certificate could be valid in AKV, but not registered as a client credential in the app registration for the app.

**Comment by wuwei8372** [comment]
@jmprieur, could we do the check to update certificateDescription.Certificate when LoadIfNeeded(CertificateDescription) is called? So that we could use CertificateDescription.KeyVaultCertificateName to find the cert in AKV even if the cert is not registered in the app registration? Then in our code we could call loadIfNeeded periodically, or just before sending each request to update certificateDescription.Certificate.

**Comment by jmprieur** [comment]
@wuwei8372, we could add a boolean to do the check. I don't think we'd want to check keyvault each  time we call LoadIfNeeded, because this would be basically each time we validate a token or acquire a token ...

**Comment by wuwei8372** [comment]
@jmprieur it makes sense to add a boolean to LoadIfNeeded to determine if checking in AKV is needed or not. I think it would work for us.

**Comment by wuwei8372** [comment]
@jmprieur could we have a cache behind LoadIfNeeded for certificate when checking certificate in AVK, so we don't have to call AKV every time if the cache is valid?

**Comment by jmprieur** [comment]
yes, that's what I mean, @wuwei8372 by LKG :)

**Comment by jmprieur** [comment]
Things to test:

Certificates pattern in ClientCertificates | In KeyVault | in app registration  | Outcome
------------------------------------------- | ------------ | --------------------- | --------------- 
1 valid certificate                                     |    Yes          |   Yes                         |  Works
1 valid certificate                                     |    No         |   Yes                          |  Log unloadable, Exception when acquiring token
1 valid certificate                                     |    Yes         |   No                          |  Exception, Exception when acquiring token
1 expired certificate                                 |    Yes         |   Yes                          | Exception when acquiring token
1 expired certificate, 1 valid certificate    |    Yes         |   Yes                          | Works
1 expired certificate, 1 valid certificate    |    Yes         |   No                          |  Exception when acquiring token

**Comment by jennyf19** [comment]
included in [1.10.0 release](https://github.com/AzureAD/microsoft-identity-web/releases/tag/1.10.0)
